File: processor.py
"""Text processing for Plus / Rage / Emoji modes.

Provider is detected from the API key prefix (same logic as transcriber.py):
  sk-...   → OpenAI  (model: gpt-4o-mini)
  gsk_...  → Groq    (model: llama-3.1-8b-instant)
"""
from __future__ import annotations
import re


def apply_snippets(text: str, snippets: list) -> str:
    """Ersetzt Keywords durch den definierten Snippet-Text (case-insensitive, ganze Wörter)."""
    for snippet in snippets:
        keyword = snippet.get("keyword", "").strip()
        replacement = snippet.get("text", "").strip()
        if not keyword or not replacement:
            continue
        pattern = re.compile(re.escape(keyword), re.IGNORECASE)
        text = pattern.sub(replacement, text)
    return text


def process(text: str, mode: str, api_key: str,
            prompt_template: str, emoji_density: int = 5,
            model: str | None = None,
            temperature: float = 0.7,
            max_tokens: int = 1024) -> str:
    """Transform *text* for the given mode. Normal mode is a no-op."""
    if mode == "normal":
        return text

    system_prompt = prompt_template
    if "{density}" in system_prompt:
        system_prompt = system_prompt.replace("{density}", str(emoji_density))

    if api_key.startswith("gsk_"):
        return _groq(text, system_prompt, api_key,
                     model or "llama-3.1-8b-instant", temperature, max_tokens)
    return _openai(text, system_prompt, api_key,
                   model or "gpt-4o-mini", temperature, max_tokens)


# ── providers ─────────────────────────────────────────────────────────

def _openai(text: str, system_prompt: str, api_key: str,
            model: str = "gpt-4o-mini", temperature: float = 0.7,
            max_tokens: int = 1024) -> str:
    import openai
    client = openai.OpenAI(api_key=api_key)
    resp = client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user",   "content": text},
        ],
        max_tokens=max_tokens,
        temperature=temperature,
    )
    return resp.choices[0].message.content.strip()


def _groq(text: str, system_prompt: str, api_key: str,
          model: str = "llama-3.1-8b-instant", temperature: float = 0.7,
          max_tokens: int = 1024) -> str:
    from groq import Groq
    client = Groq(api_key=api_key)
    resp = client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user",   "content": text},
        ],
        max_tokens=max_tokens,
        temperature=temperature,
    )
    return resp.choices[0].message.content.strip()

File: C-Sharp/Core/Processor.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Collections.Generic;

namespace Blitztext.Core;

/// <summary>
/// Processes transcribed text for Plus / Rage / Emoji modes via Chat API.
/// Normal mode is a no-op. Snippets are applied after processing.
/// Provider detected from API key prefix (sk- = OpenAI, gsk_ = Groq).
/// </summary>
public static class Processor
{
    private static readonly HttpClient _http = new()
    {
        Timeout = TimeSpan.FromSeconds(60),
    };

    public static async Task<string> ProcessAsync(
        string text,
        string mode,
        string apiKey,
        string promptTemplate,
        int emojiDensity = 5,
        string? model = null,
        float temperature = 0.7f,
        int maxTokens = 1024)
    {
        if (mode == "normal") return text;

        var prompt = promptTemplate.Replace("{density}", emojiDensity.ToString());

        return apiKey.StartsWith("gsk_")
            ? await GroqChatAsync(text, prompt, apiKey, model ?? "llama-3.1-8b-instant", temperature, maxTokens)
            : await OpenAiChatAsync(text, prompt, apiKey, model ?? "gpt-4o-mini", temperature, maxTokens);
    }

    public static string ApplySnippets(string text, IEnumerable<SnippetEntry> snippets)
    {
        foreach (var s in snippets)
        {
            if (string.IsNullOrEmpty(s.Keyword)) continue;
            text = Regex.Replace(
                text,
                Regex.Escape(s.Keyword),
                s.Text,
                RegexOptions.IgnoreCase);
        }
        return text;
    }

    // ── OpenAI ────────────────────────────────────────────────────────
    private static async Task<string> OpenAiChatAsync(
        string text, string systemPrompt, string apiKey,
        string model, float temperature, int maxTokens)
    {
        var body = BuildChatBody(text, systemPrompt, model, temperature, maxTokens);
        using var req = new HttpRequestMessage(HttpMethod.Post,
            "https://api.openai.com/v1/chat/completions");
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
        req.Content = new StringContent(body, Encoding.UTF8, "application/json");

        using var resp = await _http.SendAsync(req);
        var json = await resp.Content.ReadAsStringAsync();
        resp.EnsureSuccessStatusCode();

        return JsonNode.Parse(json)?["choices"]?[0]?["message"]?["content"]
            ?.GetValue<string>()?.Trim() ?? text;
    }

    // ── Groq ──────────────────────────────────────────────────────────
    private static async Task<string> GroqChatAsync(
        string text, string systemPrompt, string apiKey,
        string model, float temperature, int maxTokens)
    {
        var body = BuildChatBody(text, systemPrompt, model, temperature, maxTokens);
        using var req = new HttpRequestMessage(HttpMethod.Post,
            "https://api.groq.com/openai/v1/chat/completions");
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
        req.Content = new StringContent(body, Encoding.UTF8, "application/json");

        using var resp = await _http.SendAsync(req);
        var json = await resp.Content.ReadAsStringAsync();
        resp.EnsureSuccessStatusCode();

        return JsonNode.Parse(json)?["choices"]?[0]?["message"]?["content"]
            ?.GetValue<string>()?.Trim() ?? text;
    }

    // ── shared ────────────────────────────────────────────────────────
    private static string BuildChatBody(
        string userText, string systemPrompt,
        string model, float temperature, int maxTokens)
    {
        var obj = new JsonObject
        {
            ["model"]       = model,
            ["temperature"] = temperature,
            ["max_tokens"]  = maxTokens,
            ["messages"] = new JsonArray
            {
                new JsonObject { ["role"] = "system", ["content"] = systemPrompt },
                new JsonObject { ["role"] = "user",   ["content"] = userText },
            },
        };
        return obj.ToJsonString();
    }
}
